testclock/clock_test.go

```diff
@@ -170,6 +170,94 @@ func (*clockSuite) TestNewTimerReset(c *gc.C) {
 	}
 }
 
+func (*clockSuite) TestNewTimerAsyncReset(c *gc.C) {
+	t0 := time.Now()
+	clock := testclock.NewClock(t0)
+	timer := clock.NewTimer(time.Hour)
+	stop := make(chan struct{})
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-stop:
+			return
+		case t := <-timer.Chan():
+			c.Errorf("timer accidentally ticked at: %v", t)
+		case <-time.After(testing.LongWait):
+			c.Errorf("test took too long")
+		}
+	}()
+	// Just our goroutine, but we don't go so far as to trigger the wakeup.
+	clock.WaitAdvance(1*time.Minute, 10*time.Millisecond, 1)
+	// Reset shouldn't trigger a wakeup, just move when it thinks it will wake up.
+	timer.Reset(time.Hour)
+	clock.WaitAdvance(1*time.Minute, 10*time.Millisecond, 1)
+	timer.Reset(time.Minute)
+	clock.WaitAdvance(30*time.Second, 10*time.Millisecond, 1)
+	// Now tell the goroutine to stop and start another one that *does* want to
+	// wake up
+	close(stop)
```

Comment:
You need to wait for the go func to finish before moving on or there is a race with the select.

Reply:
fair point. I can't get it to fail but I do see what you mean.